apps/src/netsim/NetSimRouterNode.js

```diff
@@ -292,11 +358,140 @@ NetSimRouterNode.prototype.buildRow_ = function () {
+ * and returns them.
+ * @param {RunLoop.clock} clock
+ * @returns {NetSimMessage[]}
+ */
+NetSimRouterNode.prototype.getAndUnscheduleMessagesDueForRouting = function (clock) {
+  var dueForRouting = [];
+  this.localRoutingSchedule_ = this.localRoutingSchedule_.filter(function (entry) {
+    if (clock.time >= entry.time) {
+      dueForRouting.push(new NetSimMessage(this.shard_, entry.row));
+      return false;
+    }
+    return true;
+  }.bind(this));
+  return dueForRouting;
+};
+
+/**
+ * Examine the queue for anything that should be handled by the local
+ * simulation, that hasn't been added to the schedule yet.  Schedule it after
+ * everything that appears before it in the router queue.
+ */
+NetSimRouterNode.prototype.scheduleNewPackets = function () {
+  var row, belongsToLocalSim, notScheduled;
+  for (var i = 0; i < this.routerQueueCache_.length; i++) {
+    row = this.routerQueueCache_[i];
+    belongsToLocalSim = this.localSimulationOwnsMessageRow_(row);
+    notScheduled = !this.isScheduled(row);
+
+    if (belongsToLocalSim && notScheduled) {
+      // Add up send times of all packets in queue up to this one, and including
+      // this one, to get its send time.
+      var sendTime = this.routerQueueCache_
+          .slice(0, i + 1)
+          .reduce(function (prev, cur) {
+            return prev + this.calculateProcessingDurationForMessage_(cur);
+          }.bind(this), this.simulationTime_);
+      this.scheduleRouting(row, sendTime);
+    }
+  }
+};
+
+/**
+ * @param {messageRow} row
+ * @param {number} sendTime
+ */
+NetSimRouterNode.prototype.scheduleRouting = function (row, sendTime) {
+  this.localRoutingSchedule_.push({
+    row: row,
+    time: sendTime
+  });
+};
+
+/**
+ * @param {messageRow} messageRow
+ * @returns {boolean} TRUE if the given row is represented in the local simulation
+ *          routing schedule.
+ */
+NetSimRouterNode.prototype.isScheduled = function (messageRow) {
+  return undefined !== _.find(this.localRoutingSchedule_, function (entry) {
+    return entry.row.id === messageRow.id;
+  });
+};
+
+/**
+ * Lets the auto-DNS part of the router simulation handle its requests.
+ * For now, auto-DNS can do "batch" processing, no throughput limits.
+ * @private
+ */
+NetSimRouterNode.prototype.tickAutoDns_ = function () {
+  if (this.isAutoDnsProcessing_) {
+    return;
+  }
+
+  // Filter DNS queue down to requests the local simulation should handle.
+  var localSimDnsRequests = this.autoDnsQueue_
+      .filter(this.localSimulationOwnsMessageRow_.bind(this))
+      .map(function (row) {
+        return new NetSimMessage(this.shard_, row);
+      }.bind(this));
+
+  // If there's anything we can process, kick off the processing.
+  if (localSimDnsRequests.length > 0) {
+    this.isAutoDnsProcessing_ = true;
+    this.processAutoDnsRequests_(localSimDnsRequests, function () {
+      this.isAutoDnsProcessing_ = false;
+    }.bind(this));
+  }
```

Comment:
Yeah... I worry that "clever" and "good" would be two different things here.  I'm not going to touch this yet.